### ISSUE TITLE
NullReference Exception after updating to 5.1.3, 5.2.0, 5.2.1

### ISSUE THREAD

**Issue by msallin** [gregsdennis/json-everything#502]
### Nuget Package

JsonSchema.Net

### Package Version

5.1.3, 5.2.0, 5.2.1

### Operating System

Linux

### .Net Target (if relevant, please specify the version in the description)

.Net (5 or after)

### A clear and concise description of the bug

We upgrated from 4.1.8 to 5.1.2 and since then we get the following error.

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Json.Schema.SchemaEvaluation.FindEvaluation(Guid id)
   at Json.Schema.KeywordConstraint.BuildEvaluation(SchemaEvaluation schemaEvaluation, JsonPointer instanceLocation, JsonPointer evaluationPath)
   at Json.Schema.SchemaConstraint.BuildEvaluation(JsonNode localInstance, JsonPointer instanceLocation, JsonPointer evaluationPath, EvaluationOptions options)
   at Json.Schema.JsonSchema.Evaluate(JsonNode root, EvaluationOptions options)

Formatted by https://st.elmah.io
```

### Please add test code or steps to reproduce the behavior

We were not able to reproduce it yet. We work on it.

### What did you expect?

That it works.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by gregsdennis** [comment]
Yeah, I've unlisted that package. Use 5.1.3 instead.

**Comment by msallin** [comment]
@gregsdennis we updated to 5.1.3 and have still the same error. May you reopen?

**Comment by gregsdennis** [comment]
If you can create a reproduction, I'll reopen the issue. In the meantime, I'll have a look at the code to see if I can find any problems.

**Comment by msallin** [comment]
Ty! We will create a reproduction. It has not highest prio but we will be there in a few weeks. We now postponed the update and also implemented a lock around the schema registry (https://github.com/gregsdennis/json-everything/issues/503).

**Comment by msallin** [comment]
We found that on [line 59](https://github.com/gregsdennis/json-everything/blob/a4aae65dd856736f42dcd0b0cf3c3bc36dfcb86b/JsonSchema/SchemaEvaluation.cs#L59) the elements of KeywordEvaluations are checked for `null`. But not on [line 64](https://github.com/gregsdennis/json-everything/blob/a4aae65dd856736f42dcd0b0cf3c3bc36dfcb86b/JsonSchema/SchemaEvaluation.cs#L64). 

Not sure if this is by intention or not.

**Comment by gregsdennis** [comment]
There shouldn't need to be a null check there.  The `KeywordEvaluations` array is correctly sized using the number of keyword constraints in the schema constraint.  That array is then filled.  There should be no more null elements in the array.

What I'm interested in is a schema that produces a scenario that causes that array to not be filled.

**Comment by msallin** [comment]
Unfortunately, it isn't that easy. We consume several Kafka topics in a generic way and we load and validate a schema (the schema is embedded into NuGet packages that define the API). I looked at the logs and when the first error occurred. I created a unit test that does the same (i.e., loading the schema and validating the message). I took the same schema and the message that failed but the test pass. After the first error, all other fail as well.

I'll think about how to verify that. We could e.g. load all schema's in sequence and not start consuming before all were loaded successfully. It either fails with a specific schema or it passes (and then we know it's a race condition).

**Comment by gregsdennis** [comment]
> the schema is embedded into NuGet packages that define the API

I wonder if those packages are open source, would it be possible to grab the schema directly from the source?

**Comment by gregsdennis** [comment]
> I created a unit test that does the same (i.e., loading the schema and validating the message)

Another option might be to just serialize and print/log the schema at this point.

**Comment by msallin** [comment]
They aren't. Give us some time and we will try every schema in a test to see if we can reproduce this.
If we find that one schema is breaking the code, we will clean all sensitive info and post it here.

**Comment by edaena** [comment]
I see this issue intermittently when I use a MetaSchemas for a property. If i change the property to be a SchemaValueType.Object, I don't see the issue.
```
("foo", new JsonSchemaBuilder()
    .Type(SchemaValueType.Array)
    .Items(new JsonSchemaBuilder()
        .Type(SchemaValueType.Object)
        .Properties(
            ("bar", MetaSchemas.Core201909),
           // ("bar", new JsonSchemaBuilder().Type(SchemaValueType.Object)) // This works
```

**Comment by gregsdennis** [comment]
> I see this issue intermittently

@edaena do you run your validations in parallel, maybe?

**Comment by edaena** [comment]
> 
It is running in parallel. When I debug, upon catching the exception I inspect the value, and I do see it pointing to `meta/core`, but I'm wondering if it's just filling it later on after it threw. Not sure if this is a good way of verifying, or how do I know that the value wasn't populated later on?
<img width="850" alt="Screenshot 2023-08-28 at 1 15 07 PM" src="https://github.com/gregsdennis/json-everything/assets/1280985/6e194f5c-83ab-4a94-8174-1c7a8b055079">

**Comment by gregsdennis** [comment]
I've had some parallel issues before. I'll try setting up a parallel test to see if I can get that case to replicate.

**Comment by gregsdennis** [comment]
No, no luck reproducing, even with parallel processing.

@edaena where exactly is the stack trace pointing?  Do you get a line number?

**Comment by mdekrey** [comment]
I found this issue experiencing the same stack trace as the OP on 5.1.3, and upgrading to 5.2.0 fixed it for me. (I encountered it via Xunit tests on a .NET Standard 2.0 project.)

**Comment by msallin** [comment]
Not fixed with `5.2.1`

**Comment by gregsdennis** [comment]
@msallin it wasn't intended to be fixed with 5.2.1.  I'm still awaiting a reproduction.  Not sure what use it is calling this out.

**Comment by msallin** [comment]
Hi @gregsdennis
Sure, just wrote it because of the comment of @mdekrey. Saw some concurrency fixes and thought it's worth a try.
Still on our backlog to create a standalone repro.

**Comment by edaena** [comment]
@gregsdennis I tried to setup a separate project to repro so that I could post it, but I haven't gotten it to repro. It still repros in the other project. I'm trying to narrow down the difference. I'll post a project once I get it to repro there.

**Comment by MODUSCarstenScholling** [comment]
@gregsdennis I can provide a hopefully working repro for this race condition. I'm not 100% sure if it is json-everything or the calling app's fault, but here we go...

- It started at [microsoft/AL: AlDoc is not working with DocFx 2.70.1](https://github.com/microsoft/AL/issues/7488).
- Went over to [dotnet/docfx: 2.70.1 causes random NullReferenceException failures with custom templates.](https://github.com/dotnet/docfx/issues/9146), referencing [PR 9041](https://github.com/dotnet/docfx/pull/9041)
- @msallin was not wrong with his [comment](#issuecomment-1670776586), an element in Keyword Evaluations is null causing evaluation to be null.

See this image: ![image](https://github.com/gregsdennis/json-everything/assets/50442977/27765809-b893-402f-9383-478c9f434be3)

I'm not a friend of images, but this one shows the issue clearly ([FindEvaluation(System.Guid) (Json.Schema.SchemaEvaluation)](https://github.com/gregsdennis/json-everything/blob/0bc5b81f7a1c8b9af818cbf3fd89345dfdce3dc0/JsonSchema/SchemaEvaluation.cs#L64C65-L64C65))

But for the repro: 

- Clone or download [docfx 2.70.3](https://github.com/dotnet/docfx)
- Download null-repro-JsonSchema.zip, extract to a new folder
- Ensure docfx is in the path
- `docfx build path/to/repro/docfx.json`

```
System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=JsonSchema.Net
  StackTrace:
   at Json.Schema.SchemaEvaluation.FindEvaluation(Guid id) in json-everything\JsonSchema\SchemaEvaluation.cs:line 64
```
`docfx build path/to/repro/docfx.json --maxParallelism 1` is working correctly.

[null-repro-JsonSchema.zip](https://github.com/gregsdennis/json-everything/files/12566040/null-repro-JsonSchema.zip)

Some references regarding the origin if you need to review it:

- [Generating help with the ALDoc tool](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/help/help-aldoc-generate-help)
- [VSCode Extension containing ALDoc (Preview only)](https://marketplace.visualstudio.com/items?itemName=ms-dynamics-smb.al)

**Comment by gregsdennis** [comment]
@MODUSCarstenScholling I find it interesting that you found it happening elsewhere, but I need a repro that I can debug.  I can't debug this.

I do note that there are several issues with the meta-schema in docfx, though.  Primarily, the schema in the repro .zip points to a meta-schema that I assume is preloaded somewhere (not looking at the docfx code) against that `$schema` URI because it's not available at the URI.  (I check because I wanted to look at it.)  I found https://github.com/dotnet/docfx/blob/main/schemas/v1.0/schema.json which _might_ be the same thing, but the ID is wrong.

- There is no recognized ID because the keyword that's used is `id` and not `$id` as it should be for draft 6, which is declared by `$schema`.
- The ID URI is just the file's location in github.com, not the dotnet.github.io URI that the repro schema's `$schema` has.

**Comment by gregsdennis** [comment]
Okay.  So from [this line](https://github.com/dotnet/docfx/blob/9ec2d2c6bf8b42509d64301c120b599236a629c6/src/Docfx.Build.SchemaDriven/Validators/SchemaValidator.cs#L17), it looks like your meta-schema is actually just the JSON Schema draft 7 meta-schema.  I'm not sure why they're re-registering it under a new URI, but 🤷 I guess?

The `$id` vs `id` thing does apply to your schema as well.  I see it's currently using `id`.

**Comment by gregsdennis** [comment]
If you could find out where in the schema and where in the instance the evaluation is on that screenshot, I might be able to figure this out.  I have your schema from the repro, I just need to know what the instance (data being validated) is.

**Comment by MODUSCarstenScholling** [comment]
@gregsdennis I have no clue about the internals and schema stuff for json. I simply wanted to get us together, hopefully resulting in better software.

I have a simple working example based on the archive:

```cs
using System.Text.Json;
using Json.Schema;
using YamlDotNet.Serialization;

string schemaFile = @"C:\temp\aldoc\template\ContentTemplate\schemas\BusinessCentralApplicationObject.schema.json";
string schemaContent = File.ReadAllText(schemaFile);

SchemaValidator validator = new(schemaContent);
IEnumerable<string> files = Directory.GetFiles(@"C:\temp\aldoc\reference", "*.yml", SearchOption.AllDirectories).ToList();

files.AsParallel().Any(f => EvaluateSchema(f, validator));

static bool EvaluateSchema(string file, SchemaValidator validator)
{
    string jsonString = JsonSerializer.Serialize(new Deserializer().Deserialize(File.ReadAllText(file)));

    return validator.Validate(JsonDocument.Parse(jsonString));
}

public class SchemaValidator
{
    private readonly JsonSchema schema;

    static SchemaValidator()
    {
        SchemaRegistry.Global.Register(new("http://dotnet.github.io/docfx/schemas/v1.0/schema.json#"), MetaSchemas.Draft7);
        SchemaRegistry.Global.Register(new("https://dotnet.github.io/docfx/schemas/v1.0/schema.json#"), MetaSchemas.Draft7);
    }

    public SchemaValidator(string json): base()
    {
        schema = JsonSchema.FromText(json, new()
        {
            AllowTrailingCommas = true
        });
    }

    public bool Validate(object obj)
    {
        var json = JsonSerializer.Serialize(obj);
        var result = schema.Evaluate(JsonDocument.Parse(json), new EvaluationOptions
        {
            ValidateAgainstMetaSchema = false,
            OutputFormat = OutputFormat.List,
        });

        return result.IsValid;
    }
}
```

That's simply copy paste of what is going on in docfx. But here we have only one parallel, while docfx has two parallel loops.

**Comment by gregsdennis** [comment]
I'm sorry, I need something that looks like [this test](https://github.com/gregsdennis/json-everything/blob/master/JsonSchema.Tests/GithubTests.cs#L840) - it reproduces the problem without using outside code.

I'm not going to copy a significant portion of docfx code into my test library.

I need a schema and some data to validate.

**Comment by MODUSCarstenScholling** [comment]
Thank you, Greg! I really appreciate your commitment to this.

Sorry if my repro was not adequate for your internal testing. The same issue occurs if you take one schema, one content and run multiple created objects AsParallel().
